src/doc/rgw/multisite-reshard.md

```diff
@@ -0,0 +1,79 @@
+
+The distinction between *index layout* and *log layout* is important, because incremental sync only cares about changes to the *log layout*. Changes to the *index layout* would only effect full sync, which uses a custom RGWListBucket extension to list the objects of each index shard separately. But by changing the scope of full sync from per-bucket-shard to per-bucket and using a normal bucket listing to get all objects, we can make full sync independent of the *index layout*. And once the replication logs are moved out of the bucket index, dynamic resharding is free to change the *index layout* as much as it wants with no effect on multisite replication.
+
+## Tasks
+
+### Bucket Reshard
+
+* Modify existing state machine for bucket reshard to mutate its existing bucket instance instead of creating a new one.
+
+* Add fields for log layout. When resharding a bucket whose logs are in the index:
+    - Add a new log layout generation to the bucket instance
+    - Copy the bucket index entries into their new index layout
+    - Commit the log generation change so new entries will be written there
+    - Create a datalog entry with the new log generation
+
+### Metadata Sync
+
+* When sync fetches a bucket instance from the master zone, preserve any private fields in the local instance. Use cls_version to guarantee that we write back the most recent version of those private fields.
+
+### Data Sync
+
+* Datalog entries currently include a bucket shard number. We need to add the log generation number to these entries so we can tell which sharding layout it refers to. If we see a new generation number, that entry also implies an obligation to finish syncing all shards of prior generations.
+
+### Bucket Sync Status
+
+* Move bucket sync status to a per-bucket object, rather than having separate status per-shard.
+    - Track current log generation and an array of markers for each shard in that generation.
+* For backward compatibility, add special handling when we get ENOENT trying to read this per-bucket sync status:
+    - If the remote's oldest log layout has generation=0, read any existing per-shard sync status objects. If any are found, resume incremental sync from there.
+    - Otherwise, initialize for full sync.
+
+### Bucket Sync
+
+* Full sync uses a single bucket-wide listing to fetch all objects.
+* Incremental sync spawns a coroutine for each log shard mentioned in the datalog.
+* When incremental sync gets to the end of a log shard (ie. listing the log returns truncated=false):
+    - If we've seen a higher generation number in the datalog, flag that shard as 'resharded' in the bucket sync status.
+    - Once all shards in the current generation reach that 'resharded' state, incremental bucket sync can advance to the next generation.
```

Comment:
@cbodley This would cause that a slow syncing shard holds the rest of the bucket. Is there a way where we could move forward with new entries somehow? Note that fetching logs doesn't mean syncing the objects themselves.
Maybe split the incremental sync to two distinct units: one that is responsible for fetching the changes into local logs, and one that processes these logs and syncs the objects:
 - Fetch remote logs, but write each log entry to where it'd map in the local shard (using local layout)
 - Per local generation-shard sync will go over the local logs

That would guarantee ordering even if remote has new generation.

The resharding process can take the existing local logs and convert them to the new layout. Synchronization of it can be done by stopping sync for that bucket while reshard happens, but there are other ways to achieve that without the need to stop the sync.

Reply:
In this proposal, there will be different statuses for each unit. The remote logs status, and the local logs status.

Reply:
right. one of the unexpressed 'requirements' is that log entries for a given object name get processed in order. this generation-by-generation model seems like the simplest way to satisfy that, but your proposal provides that same guarantee

> The resharding process can take the existing local logs and convert them to the new layout.

one of the main difficulties with resharding in multisite is that our sync status markers aren't stable across the reshard of their logs. for this proposal to work, we would need some new log primitive that solved that problem where cls_log/fifo/etc cannot. without that, i feel like the only workable solution is to avoid trying to reshard the logs, and maintain distinct log generations that preserve the order without invalidating sync status markers

it's true that this generation model adds new ordering constraints across reshards, such that one shard that's behind would block progress in later shards. but given that these generations are in a strict temporal order, i think it's okay to focus exclusively on the shards that are furthest behind

this does mean that sync bandwidth is reduced across reshards, which is unfortunate. but your proposal to optimize reshards with an extra local log adds complexity and overhead to the general case, and i don't think that's the right tradeoff to make here